src/backend/distributed/commands/foreign_constraint.c

```diff
@@ -141,39 +148,56 @@ ErrorIfUnsupportedForeignConstraintExists(Relation relation, char referencingDis
 
 		/* set referenced table related variables here if table is referencing itself */
 
+		char referencedReplicationModel = REPLICATION_MODEL_INVALID;
 		if (!selfReferencingTable)
 		{
 			referencedDistMethod = PartitionMethod(referencedTableId);
 			referencedDistKey = (referencedDistMethod == DISTRIBUTE_BY_NONE) ?
 								NULL :
 								DistPartitionKey(referencedTableId);
 			referencedColocationId = TableColocationId(referencedTableId);
+			referencedReplicationModel = TableReplicationModel(referencedTableId);
 		}
 		else
 		{
 			referencedDistMethod = referencingDistMethod;
 			referencedDistKey = referencingDistKey;
 			referencedColocationId = referencingColocationId;
+			referencedReplicationModel = referencingReplicationModel;
 		}
 
-		bool referencingIsReferenceTable = (referencingDistMethod == DISTRIBUTE_BY_NONE);
-		bool referencedIsReferenceTable = (referencedDistMethod == DISTRIBUTE_BY_NONE);
+		bool referencingIsNoneDistTable = (referencingDistMethod == DISTRIBUTE_BY_NONE);
+		bool referencedIsNoneDistTable = (referencedDistMethod == DISTRIBUTE_BY_NONE);
 
-		/*
-		 * We support foreign keys between reference tables. No more checks
-		 * are necessary.
-		 */
-		if (referencingIsReferenceTable && referencedIsReferenceTable)
+		if (referencingIsNoneDistTable && referencedIsNoneDistTable)
 		{
+			bool referencingIsReferenceTable =
+				(referencingReplicationModel == REPLICATION_MODEL_2PC);
+			bool referencedIsCitusLocalTable =
+				(referencedReplicationModel != REPLICATION_MODEL_2PC);
+			if (referencingIsReferenceTable && referencedIsCitusLocalTable)
+			{
+				if (!(BehaviorIsRestrictOrNoAction(constraintForm->confdeltype) &&
```

Comment:
> Also as we only support RestrictOrNoAction in the error message we could say that only they are supported instead of writing the ones that are not supported. Because we don't have control over what is not supported and it might change, but we know what we support.

agree